### PR TITLE
Fix `MethodGenerator` parameter sorting, when an explicit sorting is provided

### PR DESCRIPTION
**Pull request by willjones9**
The MethodGenerator doesn't seem to take any notice of the ParameterGenerator position property. This causes issues when adding properties out of order, especially with optional and variadic parameters.

### CODE REVIEW

File: src/Generator/MethodGenerator.php [hunk 1]
```diff
@@ -29,6 +30,8 @@ class MethodGenerator extends AbstractMemberGenerator
 
     private bool $returnsReference = false;
 
+    private bool $sortParametersOnSet = true;
```
Comment (Ocramius) [inline src/Generator/MethodGenerator.php]: Why is this required?

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: src/Generator/MethodGenerator.php (hunk 1 above)
Comment (willjones9) [reply src/Generator/MethodGenerator.php]: `setParameters` calls `setParameter` in a loop and `setParameter` calls `sortParameters`. Having `sortParameters` called multiple times seems a little inefficient, so `$sortParametersOnSet` gets set to `false` at the start of `setParameters` to avoid needless calls.

Comment (Ocramius) [reply src/Generator/MethodGenerator.php]: @willjones-stratagem consider that `laminas/laminas-code` is not designed for runtime usage anyway: only for building sources.

I suggest removing any state mutation, and re-performing the operation, even if inefficient at first :-)

Comment (willjones9) [reply src/Generator/MethodGenerator.php]: @Ocramius fair shout. I'll update that.
That said, the project I'm working on at the moment will probably end up using `laminas/laminas-code` as part of certain runtime functionality so some overall efficiency savings in [5.x.x](https://github.com/laminas/laminas-code/issues/69) wouldn't be a bad thing

Comment (Ocramius) [reply src/Generator/MethodGenerator.php]: That is alright, but I'd like to prevent state mutation as much as possible.

An exception thrown during a flag switch would cause a mess, for example, and subclassing makes this much more complex too.

An alternate design will be needed, if this performance impact is relevant in future.

Comment (willjones9) [reply src/Generator/MethodGenerator.php]: Very good point. The PR's now been updated.

File: src/Generator/MethodGenerator.php [hunk 2]
```diff
@@ -305,6 +310,18 @@ public function setReturnsReference($returnsReference)
         return $this;
     }
 
+    /**
+     * Sort parameters by their position
+     *
+     * @return void
+     */
+    protected function sortParameters()
```
Comment (Ocramius) [inline src/Generator/MethodGenerator.php]: We can declare `: void` here.

Method can also be `private`.

File: test/Generator/MethodGeneratorTest.php [hunk 3]
```diff
@@ -74,6 +74,38 @@ public function testMethodParameterMutator()
         $methodGenerator->setParameter(new stdClass());
     }
 
+    public function testMethodParameterSetPosition()
```
Comment (Ocramius) [inline test/Generator/MethodGeneratorTest.php]: We should have two separate test methods:

 * one verifying `setParameter()`
 * one verifying `setParameters()`

File: test/Generator/MethodGeneratorTest.php [hunk 4]
```diff
@@ -74,6 +74,38 @@ public function testMethodParameterMutator()
...
+            ParameterGenerator::fromArray(['name' => 'baz', 'type' => stdClass::class, 'position' => 1]),
+            $bar
+        );
+
+        /** @var ParameterGenerator $baz */
+        $baz = array_shift($params);
+        self::assertSame('bar', $baz->getName());
+
+        $this->expectException(InvalidArgumentException::class);
+        $methodGenerator->setParameter(new stdClass());
```
Comment (Ocramius) [inline test/Generator/MethodGeneratorTest.php]: This shouldn't be needed: it's not valid from a type perspective, and does not need a runtime test.

File: test/Generator/MethodGeneratorTest.php [hunk 5]
```diff
@@ -74,6 +74,38 @@ public function testMethodParameterMutator()
...
+        $methodGenerator->setParameter(['name' => 'bar', 'type' => 'array', 'position' => 2]);
+        $methodGenerator->setParameter(
+            ParameterGenerator::fromArray(['name' => 'baz', 'type' => stdClass::class, 'position' => 1])
+        );
+
+        $params = $methodGenerator->getParameters();
+        self::assertCount(3, $params);
+
+        /** @var ParameterGenerator $foo */
+        $foo = array_shift($params);
```
Comment (Ocramius) [inline test/Generator/MethodGeneratorTest.php:88]: Suggestion:

 1. `$sorting = array_map(static function (ParameterGenerator $parameter): string { return $parameter->getName(); }, $params);`
 2. `self::assertEquals(['foo', 'bar', 'baz'], $sorting);`

Would make the test much simpler to read, IMO, and removes the need for counting returned values.

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
@willjones-stratagem I just noticed that this patch targets `4.5.x`, but this is a bugfix.

Should we change the target branch to `4.4.x`?

**Comment by willjones9** [comment]
@Ocramius retargeting it works for me

**Review by Ocramius — APPROVED** [review]
LGTM :+1:

**Comment by Ocramius** [comment]
Thanks @willjones-stratagem!
